src/components/styles/Spacing.ts

```diff
@@ -0,0 +1,75 @@
+import { css } from 'styled-components';
+import { spacing as sizes } from '../../constants/spacing';
+import grid, { TGridBreakpoints } from '../../constants/grid';
+
+type TSpacingPositionNames = 'top' | 'right' | 'bottom' | 'left' | 'x' | 'y';
+type TSpacingTypes = 'margin' | 'padding';
+type TSpacingFactoryCallback = (pos: string, index: number, position: string, size: number) => string;
+
+const positions: TSpacingPositionNames[] = ['top', 'right', 'bottom', 'left', 'x', 'y'];
+
+const spacingCssFactory = (callback: TSpacingFactoryCallback): string =>
+  sizes
+    .map((sizingValue, sizingIndex) =>
+      positions.map(position => callback(position.substr(0, 1), sizingIndex, position, sizingValue)).join(''),
+    )
+    .join(' ');
+
+const propertyFactory = (type: TSpacingTypes, sizingValue: number, position: string): string => {
+  if (position == 'y') {
+    return `
+        ${type}-top: ${sizingValue}px;
+        ${type}-bottom: ${sizingValue}px;
+      `;
+  }
+
+  if (position == 'x') {
+    return `
+        ${type}-left: ${sizingValue}px;
+        ${type}-right: ${sizingValue}px;
+      `;
+  }
+
+  return `${type}-${position}: ${sizingValue}px;`;
+};
+
+export const spacing = (type: TSpacingTypes) => css`
+
+  ${sizes
+    .map((sizingValue, sizingIndex) => `.${type.substr(0, 1)}-${sizingIndex} { ${type}: ${sizingValue}px; }`)
+    .join('')}
+
+  ${spacingCssFactory(
+    (shortHandPosition, sizingIndex, position, sizingValue) => `
+      .${type.substr(0, 1)}${shortHandPosition}-${sizingIndex} {
+        ${propertyFactory(type, sizingValue, position)}
+      }`,
+  )}
+
+  ${Object.keys(grid.breakpoints)
+    .map(
+      breakpoint => `
+      @media screen and (min-width: ${grid.breakpoints[breakpoint as TGridBreakpoints]}px) {
+
+        ${sizes
+          .map(
+            (sizingValue, sizingIndex) =>
+              `.${breakpoint as TGridBreakpoints}${CSS.escape(':')}${type.substr(
+                0,
+                1,
+              )}-${sizingIndex} { ${type}: ${sizingValue}px; }`,
+          )
+          .join('')}
+
+        ${spacingCssFactory(
+          (shortHandPosition, sizingIndex, position, sizingValue) => `
+        .${breakpoint as TGridBreakpoints}${CSS.escape(':')}${type.substr(0, 1)}${shortHandPosition}-${sizingIndex} {
+          ${propertyFactory(type, sizingValue, position)}
+        }
+      `,
+        )}
+      }
+    `,
+    )
+    .join(' ')}
+`;
```

Comment:
there is quite a bit of logic here and I think it would simplify things a bit and save us a few lines actually if we did some of the stuff by hand, i.e.
```
const createSpacingStyles = (type: 'm' | 's') =>
  Object.keys(grid.breakpoints).reduce(
    (mediaQueries, breakpoint) => ({
      ...mediaQueries,
      [`@media screen and (min-width: ${grid.breakpoints[breakpoint as TGridBreakpoints]}px)`]: {
        ...sizes.reduce(
          (classNames, size) => ({
            ...classNames,
            [`.${breakpoint}:${type}a-${size}`]: {
              margin: size,
            },
            [`.${breakpoint}:${type}v-${size}`]: {
              marginTop: size,
              marginBottom: size,
            },
            [`.${breakpoint}:${type}h-${size}`]: {
              marginRight: size,
              marginLeft: size,
            },
            [`.${breakpoint}:${type}t-${size}`]: {
              marginTop: size,
            },
            [`.${breakpoint}:${type}r-${size}`]: {
              marginRight: size,
            },
            [`.${breakpoint}:${type}b-${size}`]: {
              marginBottom: size,
            },
            [`.${breakpoint}:${type}l-${size}`]: {
              marginLeft: size,
            },
          }),
          {},
        ),
      },
    }),
    {},
  );
```

Reply:
haven't tried if it works but just to give an idea

Reply:
also it already includes the variation with spacing on all sides :)

Reply:
Thanks for the comments @thegrinder I have updated the code.

Reply:
I also swapped out a, h & v as it's closer to the tailwind documentation.